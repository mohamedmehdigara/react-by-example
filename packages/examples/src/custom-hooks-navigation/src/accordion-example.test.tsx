import {
  cleanup,
  fireEvent,
  render,
  RenderResult,
} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React, { FC, useMemo } from 'react'
import { Accordion } from './accordion-example'

describe('Accordion', () => {
  let component: RenderResult

  const Test: FC = () => {
    const items = useMemo(
      () => [
        { id: 'hello', title: 'Hello', content: <p>hello content</p> },
        { id: 'world', title: 'World', content: <p>world content</p> },
        {
          id: 'accordion',
          title: 'Accordion',
          content: <p>accordion content</p>,
        },
        { id: 'example', title: 'Example', content: <p>example content</p> },
      ],
      [],
    )
    return <Accordion items={items} />
  }

  beforeEach(() => {
    component = render(<Test />)
  })

  afterEach(() => {
    cleanup()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })

  describe('when clicking item', () => {
    beforeEach(() => {
      userEvent.click(component.getByText(/Hello/))
    })

    it('should snapshot', () => {
      expect(component.container).toMatchSnapshot()
    })

    it('should activate item', () => {
      expect(component.container).toHaveTextContent(/hello content/)
    })

    describe('when clicking item', () => {
      beforeEach(() => {
        userEvent.click(component.getByText(/Hello/))
      })

      it('should hide item', () => {
        expect(component.container).not.toHaveTextContent(/hello content/)
      })
    })

    describe('when clicking other item', () => {
      beforeEach(() => {
        userEvent.click(component.getByText(/Accordion/))
      })

      it('should show both items', () => {
        expect(component.container).toHaveTextContent(/hello content/)
        expect(component.container).toHaveTextContent(/accordion content/)
      })
    })
  })

  describe('when using keyboard navigation', () => {
    let b1: HTMLElement
    let b2: HTMLElement
    let b3: HTMLElement
    let b4: HTMLElement

    beforeEach(() => {
      ;[b1, b2, b3, b4] = component.getAllByRole('button')
    })

    describe('when pressing up', () => {
      beforeEach(() => {
        fireEvent.keyDown(b1, { key: 'ArrowUp' })
      })

      it('should focus item', () => {
        expect(b4).toHaveFocus()
      })
    })

    describe('when pressing down', () => {
      beforeEach(() => {
        fireEvent.keyDown(b2, { key: 'ArrowDown' })
      })

      it('should focus item', () => {
        expect(b3).toHaveFocus()
      })
    })
  })
})
