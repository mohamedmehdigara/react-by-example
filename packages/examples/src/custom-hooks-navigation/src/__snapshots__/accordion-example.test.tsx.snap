// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should snapshot 1`] = `
<div>
  <div>
    <div>
      <button
        class="button"
        type="button"
      >
        ðŸ”¼
         
        Hello
      </button>
    </div>
    <div>
      <button
        class="button"
        type="button"
      >
        ðŸ”¼
         
        World
      </button>
    </div>
    <div>
      <button
        class="button"
        type="button"
      >
        ðŸ”¼
         
        Accordion
      </button>
    </div>
    <div>
      <button
        class="button"
        type="button"
      >
        ðŸ”¼
         
        Example
      </button>
    </div>
  </div>
</div>
`;

exports[`Accordion when clicking item should snapshot 1`] = `
<div>
  <div>
    <div>
      <button
        class="button"
        type="button"
      >
        ðŸ”½
         
        Hello
      </button>
      <div>
        <p>
          hello content
        </p>
      </div>
    </div>
    <div>
      <button
        class="button"
        type="button"
      >
        ðŸ”¼
         
        World
      </button>
    </div>
    <div>
      <button
        class="button"
        type="button"
      >
        ðŸ”¼
         
        Accordion
      </button>
    </div>
    <div>
      <button
        class="button"
        type="button"
      >
        ðŸ”¼
         
        Example
      </button>
    </div>
  </div>
</div>
`;
